Don't let your dreams be dreams. DoIt! Just DoIt!

slaperche-scality opened the pull request:
Here comes the long awaited PR that puts an end to GNU Make felonies!

## Comparison

Here is a little timing comparison between `make` and `doit`:
- `make`: `59,41s user 4,26s system 34% cpu 3:03,44 total`
- `doit`: `56,86s user 4,53s system 33% cpu 3:03,24 total`

=> no performance issue so far :)

And in parallel:
- `doit -n 4`: `75,00s user 4,44s system 90% cpu 1:27,71 total`

I can't compare with `make -j 4` because of buggy dependencies, see #785.

## How to review

The PR is quite big but is composed of meaningful commits which should ease the process if you're reviewing commit by commit (that's how a PR is meant to be reviewed anyway).

Commit messages contains explanations and links to relevant documentation sections, but you may want to wait for my presentation tomorrow before starting the review (or you can take a look now, as you prefer)

## Conclusion

![doit](https://user-images.githubusercontent.com/34654278/54610821-1024c980-4a56-11e9-98a5-144439871f26.jpg)

slaperche-scality committed:
build: introduce doit

Our new build system will be based on `doit`[1] instead of GNU Make.
This will certainly be more maintainable.

This commit adds the necessary requirements.txt to install doit (as well
as mypy (type annotations FTW) and pylint as dev dependencies).

It also update the gitignore to ignore the cruft that we don't want
to track in Git.

[1]: http://pydoit.org/

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add configuration and global constants

Both `config.py` and `constants.py` defines global variables, but:
- those in `config.py` can be modified by the user.
- those in `constants.py` are not customisable.

It's good to keep them separated, so the people won't modify something
they're not supposed to touch (also, at one point we may replace this
`config.py` by a real configuration like `buildconf.ini` or something
similar).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/lint: add a task for linting

Lint tools can be run through `doit` now. You can either:
- run a specific linter (`doit lint:yaml`)
- run all the linter (`doit lint`)

You can list the available lint tasks with `doit list --all lint`

Each lint tool has a its own sub-task, to allow you to run only one of
them and/or run them all in parallel (`doit -n 2 lint`).

Note that we added file dependencies to the lint tasks, so that they are
only executed if the files have changed since the last lint run ;)

We also add little magic with `__all__` (only export task-creator) so
that we can safely and conveniently use wildcard import.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/vagrant: add a task to run `vagrant`

`make vagrantup` is now superseded by `doit vagrant`.

We explicitely set `uptodate`[1] to False as this task can never be
up-to-date (i.e. each time you run it you want it to spawn an
environment).

Note that this task should depend on tasks that populate the ISO, but we
can't do it yet (because those tasks aren't available).

This will be done in later commits, as we keep migrating the Makefile.

[1]: http://pydoit.org/dependencies.html#uptodate

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a thin object layer on tasks

Using task as dict is convenient and easy, but for complex task it's
likely that you'll want to combine some fields or do some computation
(e.g. infer the path where to store a package from its name and version,
…) and to do so you'll need an object.

There is also some default behavior that we'll want to apply (e.g.
allowing task to "revert" their actions, which is done by setting
`clean` to True in most of the cases) and a class is also convenient
here.

That's enough reason to create this `Target` abstract classes.

Another thing to consider is that while simple target will probably be
produced by a single task, more complex pone (such as creating a
package) will be more involved and will likely require several tasks to
be accomplished: enter `AtomicTarget` and `CompositeTarget`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/utils: add an helper for a nicer output

`title` are used to represent a task in doit's output.
Frequently we will want to display the command (or a symbolic command)
followed by its first target (especially since we will usually only have
one target in fact), e.g.:
- "MKDIR    _build/foo"
- "MKISOFS  bar.iso"
- "RENDER   baz.sls.in"
- …

Since it's a common need, it deserves an helper function.

Note that in order to have a nicer output, we need two things:
- alignment: to do so, we define a max length for the symbolic command
- avoid overlong line: absolute paths are converted to path relative to
  the build root

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a generic task to create a dir

Creating a directory will be repeated a lot, so it makes sense to have a
quick way to generate a task for that.

The advantage of having a dedicated class is two fold:
- we can hide the gory details: we just write `Mkdir(path_to_dir)` and
  the class takes care of populating `targets`, `title`, `actions`, …
  fields
- since our class inherit from `Target` and forward its `kwargs` to it,
  it's possible to customize the task at declaration time (like adding a
  `task_dep` if necessary`)

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add a task to create the build root

As it only need to be run once, we use the helper `run_once`[1].

Note the double `_`: it makes the task private.

By default, the name of the task is everything after `task_` AND tasks
whose names starts with an underscore are private/hidden so
`task__build_root` creates the task `_build_root` which is private.

Using private tasks avoid cluttering the output of `doit list` with
rarely used tasks (and if you really want to see them, you can still use
`doit list -p` or `doit list --private`).

[1]: http://pydoit.org/uptodate.html#run-once

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/iso: add a module to build the final ISO

For now we only create the root directory.

This module will grow as we continue to convert the Makefile targets to
doit tasks.

We configure `doit` to make `iso` the default task (like `iso` is the
default target in our Makefile).
And since we're addind configuration, we enable by default the cleanup
of the task dependencies: by doing that, when we simply type
`doit clean iso` to the task and its dependencies.

Without this option, we would have to type `doit clean -c iso` (
or `doit clean --clean-dep iso`), otherwise nothing would be cleaned
(because `iso` is a task-group, it produces nothing by itself and thus
has nothing to clean).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/constants: load version information

These information will be required to generate `product.txt` and
`bootstrap.sh`.

We inject them as globals, so that they can be used like the other
constants.

We can't include them directly in `config.py` because I think that the
CI relies on the VERSION file, and since we don't want to duplicate this
information then we have to import it.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add the FileTarget base class

Having a task producing a single output file is common enough (rendering
template, computing a checksum, building a container image) to deserver
a base class.

This base class does the bare minimum: setting the target to a single
element list and exposing a property (`destination`)  that avoids/hides
the ugly `targets[0]`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a generic task for template rendering

Another common needs during a build is to generate files from template
that contains placeholder that can only be filled at build time (e.g.
bootstrap.sh, top.sls, …).

To do so, we add a new generic task which is based on the standard
Python template engine (less powerful than Jinja, for sure but for now
our needs are pretty limited so there is no reason to add an
extra-dependency for that).

Note that by default, the placeholder delimiter is `$` which is a pretty
bad choice when you templatize shell script (you'll have to escape a lot
of legitimate `$`).
We will use `@@` instead (so instead of writing `@VERSION@` we will
write `@@VERSION`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/iso: add a task to populate the ISO tree

Populating the ISO tree with all the required files (container images,
software repositories, Salt states, …) will be the bulk of the work done
by our build system, so we can naturally expect this to span over
several tasks.

Now there are some cases where one could want to prepare everything for
the ISO without actually building an ISO: e.g. the task `vagrant` don't
need to the final ISO, just deploying the files is enough (they are then
"mounted" into the environment).

This commit introduce a task-group[1] that will gathers every task that
create files into the ISO file tree.

`make all-local` is superseded by the task-group `populate_iso`.

[1]: http://pydoit.org/dependencies.html?#groups

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/iso: add a task to generate the bootstrap.sh

Using a brand new helper `TemplateFile`, it is now easy to render the
bootstrap script `bootstrap.sh` from its template `bootstrap.sh.in`.

Note that this new task depends on the VERSION file, because we use the
SHORT_VERSION (which is computed from this file) as value during the
rendering, consequently if VERSION changes, we must re-render the
script.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/iso: add a task to generate product.txt

Instead of relying on an external shell script (such
`scripts/product.sh`) we generate most of the file's content from Python
itself (except for the git info), which gives us some advantages (less
fork/more portability).

We also group together the name/paths of external commands we're using:
- that way it's easy to see our depedencies
- if someone has an exotic install, they can tweak these variable as
  they need in order to make it work

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/iso: add a task to build the final ISO

Note that the dependencies computation for this task must be delayed
until the ISO root is populated (hence the use of `create_after`),
because we wants to depends on all the files in the ISO tree and this
can only be computed after populating the ISO tree.

Thanks to delayed task[1] AND ou task-group `populate_iso` (which is a
savior here, as we can only specify one single task for `executed`)
introduced earlier, it's relatively easy to do.

[1]: http://pydoit.org/task_creation.html#delayed-task-creation

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/coreutils: implement sha256sum

This commit adds a pure-Python implementation of the `sha256sum` command
(and an associated task template to ease its reuse).

The advantages of using a Python function instead of calling the
original binary are:
- portability (even though `sha256sum` is pretty ubiquitous).
- avoid a fork.
- no need to spawn a shell (`shell=True` for `subprocess.Popen`) to use
  redirections (e.g. `sha256sum foo > bar`), which is a good things to
  avoid.

Note that we take care to use incremental hashing, because loading the
whole file into memory may not be a good idea (we will checksum an ISO
and an ISO can take several Go of memory…).

Also, we can expect that the Python dev are not totally dumb that
`sha256` is in fact calling C code so the perf shouldn't be abysmal.

Anyway, let's verify it with a `hash.py` containing:

    coreutils.sha256sum([Path("_build_ref/metalk8s.iso")], Path("/tmp/sha256sum"))

We get:

    % hyperfine 'sha256sum _build_ref/metalk8s.iso > /tmp/sha256sum' './hash.py'
    Benchmark #1: sha256sum _build_ref/metalk8s.iso > /tmp/sha256sum
    Time (mean ± σ):      2.110 s ±  0.003 s    [User: 1.987 s, System: 0.121 s]
    Range (min … max):    2.106 s …  2.115 s    10 runs

    Benchmark #2: ./hash.py
    Time (mean ± σ):      2.167 s ±  0.033 s    [User: 2.016 s, System: 0.146 s]
    Range (min … max):    2.148 s …  2.258 s    10 runs

    Summary
    'sha256sum _build_ref/metalk8s.iso > /tmp/sha256sum' ran
        1.03 ± 0.02 times faster than './hash.py'

Looks pretty acceptable to me :)

And in fact, timing are even closer if you doesn't account for the
startup time of the Python interpreter.

    % python3 -m timeit -r 10 -u msec \
        -s 'from pathlib import Path' \
        -s 'from buildchain import coreutils' \
        'coreutils.sha256sum([Path("_build_ref/metalk8s.iso")], Path("/tmp/sha256sum"))'
    1 loop, best of 10: 2.11e+03 msec per loop

(2110 msec => 2.11s => same time as the `sha256sum` binary)

So yeah, "Looks pretty acceptable to me."² :p

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/iso: add a task to compute the checksum

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add a custom reporter

Most of our tasks are private/hidden, which is nice to have a clean
`doit list` output, but they are also not displayed during the execution
(so we basically display nothing, which is weird…).

Let's write our custom reporter[1] to display both public and private
tasks.

[1]: http://pydoit.org/cmd_run.html#custom-reporter

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/coreutils: implement gzip

For the same reasons we've reimplemented `sha256sum`, we provide a
Python equivalent to `gzip`.

We use 6 instead of 9 as default compression level, because that's what
the `gzip` command does (cf. man page).

Unlike `sha256sum`, we don't create a generic task for this one because
it will be used as an action inside existing task instead of being a
standalone task (note that if we ever change our mind, this will be easy
enough to add).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a ContainerImage base class

We will have to deal with two kinds of container images during our build
process:
- the existing ones that we download from registries
- the local ones that we build by ourselves

Both kinds share a common set of properties/behaviors (they both have
names, they both have tags, …) so it makes sense to have a base
ContainerImage class.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a generic task to pull images

With this task, we can pull container images from online registry and
save them (optionally compressed) locally to embed them in our ISO.

A notable change over the Makefile/shell approach is that here we're
first saving the uncompressed image on disk AND then compressing it
wherease the Makefile use on-the-fly compression (thanks to the pipe).

Is our approach harmful for the performance? Let's see, with a `zip.py`
containining:

    subprocess.run(['docker', 'save', 'salt-master:2018.3.3-1', '-o', 'img.tar'])
    coreutils.gzip(Path('img.tar'))

We get:

    % hyperfine 'docker save salt-master:2018.3.3-1 | gzip > img1.gz' './zip.py'
    Benchmark #1: docker save salt-master:2018.3.3-1 | gzip > img1.gz
    Time (mean ± σ):     18.349 s ±  0.436 s    [User: 14.851 s, System: 0.514 s]
    Range (min … max):   17.984 s … 19.179 s    10 runs

    Benchmark #2: ./zip.py
    Time (mean ± σ):     18.140 s ±  0.478 s    [User: 13.995 s, System: 0.515 s]
    Range (min … max):   17.407 s … 19.030 s    10 runs

    Summary
    './zip.py' ran
        1.01 ± 0.04 times faster than 'docker save salt-master:2018.3.3-1 | gzip > img1.gz'

Seems like they're pretty equivalent (with a lil' speed advantage for us
this time, maybe because we fork once where the shell solution fork
twice).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/image: add a module to put images on the ISO

For now we only create the root directory for the container images on
the ISO.

This module will grow as we continue to convert the Makefile targets to
doit tasks.

Because we will have several tasks (mkdir, pull, build) in this module
we create a top-level task-group to gather them. And we make this group
a public task as ones may want to build/download the images only.

Also, considering that this task will put files into the ISO tree, it
must be added as dependency to `_iso_populate`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/image: add a task to download images

Note that most images are compressed, but some aren't and it's a for a
reason.

To cite Nicolas:

    […]
    FWIW, the reason: while we can have tar.gz dumps for most images, at
    least the ones we import in the registry we deploy, we need plain
    tar for a couple of images which need to be imported in the
    containerd cache (currently using ctr cri load), which only
    understands the tar format, not tar.gz
    […]

Source: https://github.com/scality/metalk8s/issues/708#issuecomment-470265940

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a generic task to build images

With this task, we can build container images from a local Dockerfile
and (optionally) save them (compressed) on the disk.

If the image is not saved on disk, a dummy empty file is created to
in the build tree to keep track of it.

Note that for now, if you copy files into your images, you have to add
them manually as `file_dep`. In the future we may "parse" the
Dockerfile, looking for `ADD` and `COPY` statements in order to
automagically populate `file_dep`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/image: add a task to build images

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add generic task to copy a file tree

Given a list of file paths, this task will first compute and create the
required directory hierarchy and then copy all the specified files.

Note that "file path" is not limited to "file path as a string", you can
also use `FileTarget` instances (as those can easily be converted to
file path) when you want to trigger more complex operation than simply
"copy the pre-existing file" (e.g. pulling a container image, rendering
a template, …).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: workaround a cleaning "bug"

Weirdly enough it seems that when `doit` clean targets it does so from
the beginning to the end whereas I expected it to delete in reverse
order.

Consequently my nice ordering of the directories target (from the root
to the leaves) blow up when cleaning (because it starts `rmdir` from the
root, which is obviously not empty).

That being said, it's easy to workaround: just reverse the sorting order
and the iteration order at creation time.

I'll probably open an issue upstream to clarify if it's a bug or not.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/salt: add a task to deploy the Salt tree

Using a generic `FileTree` target, we create two tasks to deploy the
Salt tree (one for the `/pillar` and the other for `/salt`).

Note that we take advantage of the `FileTarget` support to render
`top.sls` and downloading the pause container image on the fly, during
the tree copy.

Like the `image` task, this one expose a public task-group and is
referenced as a depedency of `_iso_populate` as it deploys files under
the ISO root

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/package: add a module to put repo on the ISO

This is the last big chunk of work to complete the build: we need to
create the software repository on the ISO.

This last step involves several steps and this commit is the first of a
series to implement the packaging.

For now, we:
- build the container image that will be used for the packaging
  operations (downloading packages, building repositories, …).
- create the root directories for the software packages and repositories

Tough work will be done in following commits.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a generic task to build packages

This target abstract all the gory details involved in package creation:
- create the directory for the package
- generate a .meta from the .spec
- download the source files using the URLs in the .meta
- build a SRPM containing the source files

We only build a SRPM here, the RPM depends on the repository and thus
will be built when we build the repository.

Because we manually declare the source file (instead of relying on the
content of the .meta file) we check that every declared source file has
an entry in the .meta file and bail out if it's not the case.
(BTW, note that `curl` is advantagely replaced by `urllib.request` :-))

Also, since all of the command using the build container share a common
set of options (like mounting the entrypoint script :p), we define them
in `constants.py` to avoid duplication.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/package: add a task to build custom packages

For now we only build Calico CNI plugin.

Packages are declared in a dict, so that they can be associated with the
repository where they belong.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/package: add a task to download packages

This will also build a repository with those packages included.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/target: add a generic task to build repo

This target abstract all the gory details involved in repository
creation.

We have two cases:
- repository that contains already existing packages
- repository that contains locally built packages

In the first case, the bulk of the work was already done by the
`_download_packages` task (it has already built the repositories around
the packages we downloaded): we only have to generate the `repomd.xml`.

In the second case, we first have to create the directories for the
repository, then build a RPM from a SRPM for each packages and finally
we can generate the `repomd.xml`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/package: add a task to download packages

"Standard" repositories depends on packages downloading whereas "custom"
repositories depends on their packages list.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add more info about doit in BUILDING.md

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add the TaskDict type

It's more explicit that simply `dict` and make the distinction with
other `Dict[str, Any]`.
Also, when mypy is configured to be more pedantic it won't accept
`dict`.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

NicolasT commented on Makefile:
```diff
@@ -214,7 +214,7 @@ all-local: $(ALL) ## Build all artifacts in the build tree
 	mkdir -p $(shell dirname $@)
 	rm -f $@
-	sed s/@VERSION@/$(shell source $(ISO_ROOT)/product.txt && echo $$SHORT_VERSION)/g < $< > $@ || (rm -f $@; false)
+	sed s/§VERSION/$(shell source $(ISO_ROOT)/product.txt && echo $$SHORT_VERSION)/g < $< > $@ || (rm -f $@; false)
```
Huh?

slaperche-scality replied:
See commit message https://github.com/scality/metalk8s/pull/795/commits/87564c9d38803d1fc713af79152ae410d225ca5d

I could have used `@{placeholder}` but then `shellcheck` complains.

NicolasT replied:
I, for one, don't have it on my keyboard, which makes editing a bit cumbersome, at least...

slaperche-scality replied:
Then it's annoying indeed. I'll try to find something in the ASCII charset that won't conflict easily and won't trigger `shellcheck`.

slaperche-scality replied:
It seems that I can use a string as delimiter (for some reason I thought that I was limited to a single char) so let's use `@@VERSION` instead of `@VERSION@`.

(thanks @gdemonet for opening my eyes)

NicolasT commented on buildchain/config.py:
```diff
@@ -0,0 +1,40 @@
+BUILD_ROOT : Path = ROOT/'_build'
+
+# Vagrant configuration.
+VAGRANT_PROVIDER : str = 'virtualbox'
```
Is there a way to override this value without changing any Git-tracked files?

slaperche-scality replied:
Not yet (same as the Makefile approach).

That's why I suggested/asked about supporting a `build.ini` configuration file that could be used for that purpose.
Seeing this comment I'll consider that we want to go that way and create an issue to track this task (or maybe do it in this PR if it's small enough).

NicolasT replied:
```make
VAGRANT_DEFAULT_PROVIDER = virtualbox
VAGRANT_UP_OPTS ?= --provision --no-destroy-on-error --parallel --proider=$(VAGRANT_DEFAULT_PROVIDER)
```

so either `make vagrantup VAGRANT_DEFAULT_PROVIDER=libvirt` or an environment variable would work ;-) Could argue the variable name could be improved upon.

NicolasT replied:
Not overly fond of some `config.ini` file. That's hard to keep in sync, or to set up in CI.

slaperche-scality replied:
Can you detail a little bit? What kind of sync issues?

Other than a config file, what could we use? Environment variables (maybe easier to use in CI, less so for a human user though)? Something even better :-)?

NicolasT replied:
> Can you detail a little bit? What kind of sync issues?

Do you check-in a config `config.ini`, which would then confict when people make local modifications? A `config.ini.example`? ...

For humans, one can use `dotenv` or similar :)

slaperche-scality replied:
> Do you check-in a config `config.ini`, which would then confict when people make local modifications? A `config.ini.example`? ...

That was the idea, yeah (a `config.ini.example`).

Is something wrong with that?

Didn't know about `dotenv`, looks nice but introduces a new dependency it seems. What's the avantage over a `.ini`?

slaperche-scality replied:
Done, with a `.env` at the root of the repo.

NicolasT commented on buildchain/constants.py:
```diff
@@ -0,0 +1,88 @@
+            name, _, value = line.strip().partition('=')
+            # Don't overwrite random variables by trusting an external file.
+            if name in to_update:
+                globals()[name.strip()] = value.strip()
```
Wouldn't it be nicer if this returns a tuple, which gets then bound in the local scope?

```python
def load_version_information():
    ...
    return (major, minor, patch, suffix)

VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH, VERSION_SUFFIX = load_version_information()
```

slaperche-scality replied:
Not sure it's nice, the `...` will probably be more complicated than the actual `…`.

So far I'm getting

```
def load_version_information() -> Tuple[str, str, str, str]:
    """Load version information from `VERSION`."""
    to_update = {
            'VERSION_MAJOR': '',
            'VERSION_MINOR': '',
            'VERSION_PATCH': '',
            'VERSION_SUFFIX': '',
    }
    with VERSION_FILE.open('r', encoding='utf-8') as fp:
        for line in fp:
            name, _, value = line.strip().partition('=')
            # Don't overwrite random variables by trusting an external file.
            if name.strip() in to_update:
                to_update[name.strip()] = value.strip()
        return (
            to_update['VERSION_MAJOR'],
            to_update['VERSION_MINOR'],
            to_update['VERSION_PATCH'],
            to_update['VERSION_SUFFIX']
        )

VERSION_FILE : Path= ROOT/'VERSION'

# Metalk8s version.
VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH, VERSION_SUFFIX = load_version_information()
```

Not convinced by the gain (but maybe there is a simpler body for `load_version_information`).

NicolasT commented on buildchain/lint.py:
```diff
@@ -0,0 +1,68 @@
+# coding: utf-8
+
+
+"""Tasks for the linting.
```
Not sure this should be part of the 'buildchain' (neither of the current `make` infrastructure, actually). This is all driven by `tox`, let's document the invocation as such.

slaperche-scality replied:
The problem with `tox` is that it seems that you can't have a "target" that run all the linter (do you confirm @jbertran?) so you would have to type `tox -e lint-shell` and `tox -e lint-yaml` (and soon `tox -e lint-python`, …).

jbertran replied:
tox has no concept of hierarchy, only targets attached to an environment, so you can't specify a "meta-target" in such a way that, for example, the lint target would trigger the lint-yaml target _and_ the lint-shell target.

This is why in the current iteration, the meta-target is in the Makefile. You can absolutely copy-paste and mash the two targets together in something that would be a meta-target in name only, but it's very ugly, and it would run the target once, as well as running its "subtargets" once, if you just ran `tox -e test`.

slaperche-scality replied:
`doit` being a task automation tool more than a build automation tool, I'm not disturbed by the `lint` target (`vagrantup` doesn't build anything as well if we go that way).

But if there is a clean `tox`-only solution I'm not against it :slightly_smiling_face:

NicolasT replied:
`vagrantup` doesn't build immediately, but does depend on `all-local`.

slaperche-scality replied:
> `doit vagrant` doesn't work in my setup because `_build/root/bootstrap.sh` is not executable.

Good catch!

There is a bug here (I should preserve the permissions/owners/… of the template when I render it).

(I mainly tested the build of the ISO, so I missed that one).

slaperche-scality replied:
Fixed (about the bootstrap.sh non-executable)

slaperche-scality replied:
I'll keep the target as-is for now (as it was present in the original Makefile and I'm porting what currently exists).

If we really want it, we can move it out in a later PR.

slaperche-scality commented:
Some things I forgot to mention:
- The code successfully pass `mypy` in strict mode and `pylint` with my especially pedantic configuration.
- I require Python 3.6 because I use `pathlib` and since Python 3.6 there is an interface, `os.PathLike`, that makes using `Path` easier.

And now, some questions for you, dear reviewers:

## config.py

We may want to replace `config.py` by a `config.ini` since it contains values that can be configured by the user.

Maybe not a priority though.

## Dockerfile

For the Docker images, we add files included in the image explicitly.
We may prefer to "parse" the Dockerfile to extract the `ADD` and `COPY` and either:
- auto populate the dependencies from this
- check against the manually specified dependencies and bail out if there is a mismatch.

Opinions?

## nginx

In the original Makefile, there was no registry (`nginx@sha256:LONG_SHA_HERE`)which i found weird so I took the liberty to add one.
I don't know if it's a mistake or not. Any thoughts?

## product.txt

Do we really want to include the build timestamp (because of that, the ISO is never up-to-date)?
Isn't the SHA1 enough to identify the build?

NicolasT commented on buildchain/targets/package.py:
```diff
@@ -0,0 +1,290 @@
+        mount_string = 'type=bind,source={src},destination={dst},ro'.format(
+            src=self.spec, dst=container_spec
+        )
+        rpmspec_cmd = [
```
Would it make sense to use the Docker Python API instead of always shell'ing out to Docker CLI?

slaperche-scality replied:
It could yeah.

My goal here was to:
- port the Makefile
- keep the build dependencies to a minimum (ideally only `doit`)

So I kept the CLI approach. But it may be worth/easier to use the Docker Python API.
If we want to go that way, i'll create a dedicated issue :-)

slaperche-scality replied:
I created an issue to keep track of that suggestion.

NicolasT commented on dodo.py:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
```
What's this file for, exactly?

slaperche-scality replied:
That's  to `doit` what `Makefile` is to `make`: your entry point

NicolasT commented on packages/packages.list:
```diff
@@ -0,0 +1,12 @@
+ftp://ftp.scientificlinux.org/linux/scientific/7x/external_products/extras/x86_64/container-selinux-2.77-1.el7_6.noarch.rpm
```
Can we keep this sorted alphabetically?

slaperche-scality replied:
Done

NicolasT commented on pillar/top.sls.in:
```diff
@@ -1,3 +1,3 @@
-metalk8s-@VERSION@:
+metalk8s-§VERSION:
```
Why?

slaperche-scality replied:
Replaced by `@@VERSION` now, which is more "portable" from a keyboard point of view.

NicolasT commented on requirements/build-dev-requirements.in:
```diff
@@ -0,0 +1,3 @@
+mypy == 0.670
```
Instead of introducing an artificial top-level `requirements` directory, let's keep requirement definition files per 'project'?

slaperche-scality replied:
Do you mean moving those files in the `buildchain` directory? ~(would work, a bit dirty as `buildchain is a Python module and that's no place to put such files (and adding a extra directory level will make weird import), but would work).~
Edit: Nevermind, I find a clean and simple way to do it, so if you meant that I can do it.

Or did I misunderstand your suggestion?

slaperche-scality replied:
Done (move the requirements.txt in the `buildchain` directory)

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -53,7 +53,7 @@ stop_salt_minion_service() {
 
 configure_yum_repositories() {
-    local -r repositories_pillar_path=/srv/scality/metalk8s-@VERSION@/pillar/repositories.sls
+    local -r repositories_pillar_path=/srv/scality/metalk8s-§VERSION/pillar/repositories.sls
```
Etc.

slaperche-scality replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented:
> product.txt
> Do we really want to include the build timestamp (because of that, the ISO is never up-to-date)?
Isn't the SHA1 enough to identify the build?

No. The SHA1 doesn't specify what's on the ISO, since we rely on retrieving content of upstream repositories, which we don't control.

NicolasT commented on Makefile:
```diff
@@ -214,7 +214,7 @@ all-local: $(ALL) ## Build all artifacts in the build tree
 $(ISO_ROOT)/bootstrap.sh: scripts/bootstrap.sh.in $(ISO_ROOT)/product.txt
```
If this PR lands, then I want the whole `make` infrastructure to be removed in one go, so we never have two systems in `development/*` at the same time.

However.

I think it may make sense to have some `PHONY` targets in a `Makefile` so `make`, `make vagrantup` etc. still 'work'.

slaperche-scality replied:
Yes, agreed.

We plan to remove it in one shot, but we need to update the CI to rely on doit instead of Make so I have to keep it for now.

Edit: Ha yes I could add a transient make that forward command to doit in this PR, then I don't have to wait for the CI update.
Could be an option yes.

NicolasT replied:
As far as I'm concerned it's not even 'transient' ;-)

slaperche-scality replied:
But then if I follow your other comment (using `tox`) we will have `make` that calls `tox` that calls `doit`: why so many levels?

For `tox` I get it: it handles the `venv` business, but what's the added value of the make layer/indirection?

NicolasT replied:
`make -n4` is quite a bit faster to type than `tox -e doit -- -n 4` (or whatever target name we choose etc).

slaperche-scality replied:
Indeed, so it may be interesting.

slaperche-scality replied:
Forwarding command line argument from `make` to underlying command is either hackish or more verbose, but I found a clean way to do that (handle the venv stuff automatically and run doit transparently).

NicolasT commented on BUILDING.md:
````diff
@@ -0,0 +1,136 @@
+## Setup
+
+```shell
+python3 -m venv buildenv
````
Can we set up a `tox` target for this, which managed all of this? So we can run `tox -e build` (~ `doit`), `tox -e build -- list` (~ `doit list`) etc?

slaperche-scality replied:
Solved with a wrapper script so that the users don't have to deal with the venv themselves.

slaperche-scality committed:
build: use relative import

This avoid cyclical import issues in Python < 3.7 (and anyway it's
cleaner).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Conflict




There is a conflict between your branch `feature/708-don_t_let_your_dreams_be_dreams` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`feature/708-don_t_let_your_dreams_be_dreams`).

```bash
 $ git fetch
 $ git checkout origin/feature/708-don_t_let_your_dreams_be_dreams
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/708-don_t_let_your_dreams_be_dreams
```

NicolasT commented:
`doit vagrant` doesn't work in my setup because `_build/root/bootstrap.sh` is not executable.

NicolasT reviewed (changes requested):
https://github.com/scality/metalk8s/pull/795#issuecomment-474496964

Also, I'd like the target name to rename `vagrantup` not `vagrant`.

NicolasT reviewed (changes requested):
Looks like the `all-local` target is gone?

slaperche-scality commented:
> Looks like the `all-local` target is gone?

Nope, see https://github.com/scality/metalk8s/pull/795/commits/0cfff351e9d40a7467b7e34bc681e9f8fd5a2e8c

It's just hidden (and renamed) because I though it was never used directly (you run `vagrantup` or `iso`). But if I was wrong I can make the task public, no problem.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented:
> > Looks like the `all-local` target is gone?
> 
> Nope, see [0cfff35](https://github.com/scality/metalk8s/commit/0cfff351e9d40a7467b7e34bc681e9f8fd5a2e8c)
> 
> It's just hidden (and renamed) because I though it was never used directly (you run `vagrantup` or `iso`). But if I was wrong I can make the task public, no problem.

That task is most definitely useful for humans (devs). It's significantly faster than `iso`, and doesn't necessarily need `vagrantup`: once a VM is up, starting `bootstrap.sh` after `all-local` inside it is quite a bit faster than running `vagrantup` which does checks against the VM first and whatnot.

Furthermore, it's not about `bootstrap.sh` only. Wile developing a new state, module,... and the Vagrant VM is already up, it's most certainly useful to be able to edit sources on your host, `make all-local` and then issue whatever `salt-call`, `crictl exec ... salt-run ...` or whatever in the VM through a `vagrant ssh` shell to test your changes without the need to run the whole bootstrap (takes time) or even integrating your changes in the bootstrap process.

We most certainly want/need such target to be available for development purposes (there's a reason it was added in the first place...)

NicolasT commented on README.md:
````diff
@@ -22,7 +22,7 @@ If you'd like to contribute, please review the
 ```shell
 # Bootstrap a platform on a vagrant environment using
-make vagrantup
+doit vagrantup
````
That won't work. While we can somewhat reasonably assume one has `make` available on a system, or can easily get it installed, chances are slim `doit` is pre-installed.

Either setting up the whole venv etc. could be added here (which is meh, and we'll run into issues where people created the venv, we update requirements, people don't re-run `pip install ...`, and things break), or we have `make` targets to set up the venv, or we introduce the `tox` based targets (wrapped by `make` `PHONY` targets?), or ...

slaperche-scality replied:
Yeah, shouldn't have changed that as we will still use a dummy Make to have short commands.

slaperche-scality replied:
Done

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/708-don_t_let_your_dreams_be_dreams` into
`development/2.0`
* [w/2.1/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.1` (pull request #806)
* [w/2.2/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.2` (pull request #807)



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
build: move everything build-related to buildchain

This avoid cluttering the top level directory.

Note that this change has bring out a little bug where we weren't using
absolute path.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build: add a wrapper script around `doit`

In order to avoid bothering the user with spawning a virtual environment
(and having to maintain it up-to-date with our requirements.txt, which
is error prone!) we introduce a minimalist shell script wrapper that
installs `doit` in a virtualenv, taking care to update it if necessary,
and then invokes `doit` with all the command line arguments passed
through (it's like using `doit` directly, which is quite convenient).

We also update the "documentation" accordingly.

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
build: add support for a .env file

Users can now customize the build variables by using a .env file.

Available options are now documented in BUILDING.md

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
> We most certainly want/need such target to be available for development purposes (there's a reason it was added in the first place...)

Totally makes sense, thanks for the explanation: `iso_populate` is now public.

slaperche-scality committed:
build: kill the Makefile

Come, thou forest, with thy heroes,
Come, ye junipers and pine-trees,
With your messengers of power,
Come, ye mountains, with your wood-nymphs,
Come, ye lakes, with all your mermaids,
Come, ye hundred ocean-spearmen,
Come, torment this son of Hisi,
Come and kill this evil monster.

Closes: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #24850215d47aa45c7d6da94763826f37562cbe0d on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #c0e8ca5d204d706b7a4ec8b89d2d5f82b48475a8 on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/708-don_t_let_your_dreams_be_dreams` into
`development/2.0`
* [w/2.1/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.1` 
* [w/2.2/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

thomasdanan commented:
> `doit vagrant` doesn't work in my setup because `_build/root/bootstrap.sh` is not executable.

Same for me ... using vagrant up ... bootstrap.sh not executable

bert-e commented:
# History mismatch



Merge commit #8d6f24ebb4c28ae5a48ed7d6c528fd668b4f13b0 on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality committed:
build/targets: fix images' tags for containerd

Images loaded into containerd have more constraints, such as no
compression and also a different format for the tag name.

Instead of adding yet another parameters to control that, let's have a
single boolean to indicate if the image should be loadable in containerd
(and if yes, we do what's necessary).

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #5be859d0050c45e35eb15eab7777f09e799ba1f5 on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality committed:
build: add linting for the buildchain code

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #e055e204ddcab5e852aff921dbac91b72303c574 on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented on README.md:
```diff
@@ -4,15 +4,21 @@
 
+## Building
+
+To build a MetalK8s ISO, simply type `./doit.sh`.
```
So, no `make` targets? :disappointed: :-P

slaperche-scality committed:
build/image: add build task for the UI image

Refs: #708
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/708-don_t_let_your_dreams_be_dreams` into
`development/2.0`
* [w/2.1/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.1` 
* [w/2.2/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# History mismatch



Merge commit #8072a3e72a6606ff66bbad648c3f35af000e6d9a on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #64f72d9d87025b45424641f1c2fbefb68d8a9d46 on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #75a9a273883563350061d91cf4da4ca30ff4da24 on the integration branch
`w/2.1/feature/708-don_t_let_your_dreams_be_dreams` is merging a branch which is neither the current
branch `feature/708-don_t_let_your_dreams_be_dreams` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/708-don_t_let_your_dreams_be_dreams` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/708-don_t_let_your_dreams_be_dreams` with contents from `feature/708-don_t_let_your_dreams_be_dreams`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/708-don_t_let_your_dreams_be_dreams origin/development/2.1
 $ git merge origin/feature/708-don_t_let_your_dreams_be_dreams
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/708-don_t_let_your_dreams_be_dreams
```

slaperche-scality commented:
Last (hopefully) rebase, to add `'salt/metalk8s/kubeadm/init/control-plane/files/htpasswd'` that was merged recently (basic auth PR)

slaperche-scality commented:
@bert-e approve

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/708-don_t_let_your_dreams_be_dreams` with contents from `feature/708-don_t_let_your_dreams_be_dreams`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/708-don_t_let_your_dreams_be_dreams origin/development/2.1
 $ git merge origin/feature/708-don_t_let_your_dreams_be_dreams
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/708-don_t_let_your_dreams_be_dreams
```





*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/708-don_t_let_your_dreams_be_dreams` into
`development/2.0`
* [w/2.1/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.1` 
* [w/2.2/feature/708-don_t_let_your_dreams_be_dreams](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/708-don_t_let_your_dreams_be_dreams) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
